mayastor/src/persistent_store.rs

```diff
@@ -0,0 +1,250 @@
+//! The persistent store is used to save information that is required by
+//! Mayastor across restarts.
+//!
+//! etcd is used as the backing store and is interacted with through the use of
+//! the etcd-client crate. This crate has a dependency on the tokio async
+//! runtime.
+use crate::{
+    core,
+    store::{
+        etcd::Etcd,
+        store_defs::{
+            DeleteWait,
+            GetWait,
+            PutWait,
+            Store,
+            StoreError,
+            StoreKey,
+            StoreValue,
+        },
+    },
+};
+use futures::channel::oneshot;
+use once_cell::sync::OnceCell;
+use serde_json::Value;
+use snafu::ResultExt;
+use std::{future::Future, sync::Mutex, time::Duration};
+
+static DEFAULT_PORT: &str = "2379";
+static STORE_OP_TIMEOUT: Duration = Duration::from_secs(30);
+static PERSISTENT_STORE: OnceCell<Option<Mutex<PersistentStore>>> =
+    OnceCell::new();
+
+/// Persistent store
+pub struct PersistentStore {
+    /// Backing store used for persistence.
+    store: Etcd,
+    /// Endpoint of the backing store.
+    endpoint: String,
+}
+
+impl PersistentStore {
+    /// Initialise the persistent store.
+    /// If the supplied endpoint is 'None', the store is uninitalised and
+    /// unavailable for use.
+    pub async fn init(endpoint: Option<String>) {
+        if endpoint.is_none() {
+            // No endpoint means no persistent store.
+            warn!("Persistent store not initialised");
+            return;
+        }
+
+        assert!(endpoint.is_some());
+
+        // An endpoint has been provided, initialise the persistent store.
+        let endpoint = Self::format_endpoint(&endpoint.unwrap());
+        let store = Self::connect_to_backing_store(&endpoint.clone()).await;
```

Comment:
Are there any other tuneables here like connection timeouts, etc ? Would it be reasonable to introduce something tuneable here to have better control over ETCd backed ?

Reply:
Ah, I see later on that we try to eternally connect to ETCd. So please disregard this comment.